src/scopeck.rs

```diff
@@ -1055,6 +1064,73 @@ pub(crate) fn scope_check(result: &mut ScopeResult, segments: &SegmentSet, names
     }
 }
 
+/// A [`Frame`] reference in the context of a [`Database`].
+/// This allows the values in the [`Frame`] to be resolved,
+#[derive(Copy, Clone, Debug)]
+pub struct FrameRef<'a> {
+    db: &'a Database,
+    frame: &'a Frame,
+}
+
+impl std::ops::Deref for FrameRef<'_> {
+    type Target = Frame;
+
+    fn deref(&self) -> &Self::Target {
+        self.frame
+    }
+}
+
+impl FrameRef<'_> {
+    /// Iterates over the essential hypotheses for this frame.
+    pub fn essentials(&self) -> impl Iterator<Item = (Label, Formula)> + '_ {
+        self.frame.hypotheses.iter().filter_map(move |hyp| {
+            if let Hyp::Essential(sa, _) = hyp {
+                let sref = self.db.parse_result().statement(*sa);
+                let label = self
+                    .db
+                    .name_result()
+                    .lookup_label(sref.label())
+                    .map_or(Label::default(), |l| l.atom);
+                let formula = self
+                    .db
+                    .stmt_parse_result()
+                    .get_formula(&sref)
+                    .map(Formula::clone)?;
+                Some((label, formula))
+            } else {
+                None
+            }
+        })
+    }
+
+    /// Iterates over the floating hypotheses for this frame.
+    pub fn floating(&self) -> impl Iterator<Item = Label> + '_ {
```

Comment:
This should be factored in two, where one iterator does the filter-mapping and returns `StatementAddress`, and the db-aware part looks up the label (or should we stop at the `StatementRef`?)

Reply:
Yes, but I'm still not sure which one to choose between the two. 
I've usually used `StatementRef` more on the API, as the information I need are usually only obtained through it.

Reply:
There is no harm in having more than one version of the iterator if a certain chain is common. The thing to avoid here is an iterator that does more work than the user needs, and I can imagine that if you are throwing the addresses in a data structure then you aren't interested in looking them up just yet, hence the bare address iterator is useful. Also, the address iterator doesn't need the DB so it can work on a `&Frame`, not a `FrameRef`.

Reply:
Ok, I did the split with `StatementAddress`.